# Team Project 3
## Hannah Garrett & Leighann Robinson
### DS160 SP 23

## Import the libraries

In [1]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

## Import the dataset

In [2]:
dataset=pd.read_csv('Housing.csv')

## View the datatset

In [3]:
dataset.head()

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


##### There are no null values to impute for this dataset.

### Splitting the dataset into independent and dependent variables

In [5]:
dataset.columns

Index(['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'furnishingstatus'],
      dtype='object')

In [6]:
X=dataset[['area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'furnishingstatus']]
y=dataset[['price']]

### Handling categorical data

In [7]:
X=pd.get_dummies(dataset[['area',
                          'bedrooms',
                          'bathrooms',
                          'stories',
                          'mainroad',
                          'guestroom',
                          'basement',
                          'hotwaterheating',
                          'airconditioning',
                          'parking',
                          'prefarea',
                          'furnishingstatus']],
                drop_first=True)

In [8]:
X.sample(10)

area  bedrooms  bathrooms  stories  parking  mainroad_yes  guestroom_yes  \
491  2640         2          1        1        1             0              0   
475  3000         2          1        2        0             1              0   
429  4775         4          1        2        0             1              0   
152  5400         5          1        2        0             1              1   
375  5960         3          1        2        0             1              1   
52   6000         4          2        4        1             1              1   
320  4320         3          1        2        2             1              0   
317  4992         3          2        2        2             1              0   
94   6000         4          2        4        1             1              0   
399  7350         2          1        1        1             1              0   

     basement_yes  hotwaterheating_yes  airconditioning_yes  prefarea_yes  \
491             0                    0                    0             0   
475             0                    0                    0             0   
429             0                    0                    0             0   
152             1                    0                    1             1   
375             1                    0                    0             0   
52              0                    0                    1             0   
320             0                    0                    0             1   
317             0                    0                    0             0   
94              0                    0                    0             0   
399             0                    0                    0             0   

     furnishingstatus_semi-furnished  furnishingstatus_unfurnished  
491                                0                             0  
475                                1                             0  
429                                0                             1  
152                                0                             0  
375                                0                             1  
52                                 0                             0  
320                                0                             0  
317                                0                             1  
94                                 1                             0  
399                                1                             0

#### Encoding key 

##### -  mainroad_yes = 1 , mainroad_no = 0 (dropped)
##### - guestroom_yes = 1 ,  guestroom_no = 0 (dropped)
##### - basement_yes = 1 , basement_no = 0 (dropped)
##### - hotwaterheating_yes = 1 , hotwaterheating_no = 0 (dropped)
##### - airconditioning_yes = 1 , airconditioning_no = 0 (dropped) 
##### - prefarea_yes = 1 , prefarea_no = 0 (dropped)
##### - furnishingstatus_semi-furnished = 10 , furnishingstatus_unfurnished = 01 , furnishingstatus_furnished = 00 (dropped)

### Splitting dataset into training and test sets

In [9]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,
                                              test_size=.2,
                                              random_state=20)

### Training MLR model on training set

In [10]:
from sklearn.linear_model import LinearRegression
regressor=LinearRegression()
regressor.fit(X_train.values, y_train)

LinearRegression()

### Finding intercept & coefficients

In [11]:
print("Coefficients: ", regressor.coef_)
print("Intercept: ", regressor.intercept_)

Coefficients:  [[ 2.42860234e+02  1.52680838e+05  9.81458164e+05  4.15303525e+05
   3.17368422e+05  4.05723366e+05  4.10717450e+05  3.29333078e+05
   4.30937815e+05  8.19728625e+05  6.75717105e+05  2.70974975e+04
  -3.33891104e+05]]
Intercept:  [-68421.47276616]


#### Equation 

In [12]:
X_train.columns

Index(['area', 'bedrooms', 'bathrooms', 'stories', 'parking', 'mainroad_yes',
       'guestroom_yes', 'basement_yes', 'hotwaterheating_yes',
       'airconditioning_yes', 'prefarea_yes',
       'furnishingstatus_semi-furnished', 'furnishingstatus_unfurnished'],
      dtype='object')

##### Price = -68421.4728 + area(*2.4286x10^2) + bedrooms(*1.5268x10^5) + bathrooms(*9.8146x10^5) + stories(*4.1530x10^5) + parking(*3.1737x10^5) + mainroad_yes(*4.0572x10^5) + guestroom_yes(*4.1072x10^5) + basement_yes(*3.2933x10^5) + hotwaterheating_yes(*4.309x10^5) + airconditioning_yes(*8.1973x10^5) + prefarea_yes(*6.7572x10^5) + furnishingstatus_semi-furnished(*2.7097x010^4) - furnishingstatus_unfurnished(*3.3389x10^5)

### Predicting test results

In [13]:
y_pred=regressor.predict(X_test.values)

### RMSE and R Square

In [14]:
from sklearn.metrics import mean_squared_error, r2_score
import math
print(f"R Square: {r2_score(y_test,y_pred):.2f}")
print(f"MSE: {mean_squared_error(y_test,y_pred):.2f}")
print(f"RMSE: {math.sqrt(mean_squared_error(y_test,y_pred)):.2f}")

R Square: 0.68
MSE: 1199438721041.63
RMSE: 1095188.90


### Validation

In [16]:
X_test.head(10)

area  bedrooms  bathrooms  stories  parking  mainroad_yes  \
56   11440         4          1        2        1             1   
290   2610         4          3        2        0             0   
16    6600         4          2        2        1             1   
410   3850         3          1        1        0             1   
129  11460         3          1        3        2             1   
461   4960         2          1        1        0             1   
77    6500         3          2        3        0             1   
248   4100         2          2        1        0             1   
281   4500         2          1        1        2             1   
172   8400         3          1        2        2             1   

     guestroom_yes  basement_yes  hotwaterheating_yes  airconditioning_yes  \
56               0             1                    0                    0   
290              0             0                    0                    0   
16               1             1                    0                    1   
410              0             0                    0                    0   
129              0             0                    0                    0   
461              0             1                    0                    1   
77               0             0                    0                    1   
248              1             1                    0                    0   
281              0             0                    0                    1   
172              1             1                    0                    1   

     prefarea_yes  furnishingstatus_semi-furnished  \
56              1                                1   
290             0                                1   
16              1                                0   
410             0                                0   
129             1                                1   
461             0                                0   
77              1                                0   
248             0                                1   
281             0                                0   
172             1                                0   

     furnishingstatus_unfurnished  
56                              0  
290                             0  
16                              1  
410                             1  
129                             0  
461                             1  
77                              0  
248                             0  
281                             0  
172                             1

#### Predict the prices for the scenarios below taken from the csv file. The actual values are  indicated below to showcase the different between the model's predictions and the actual price. 

### backwards eliination?